resources/user.go

```diff
@@ -0,0 +1,212 @@
+// Mgmt
+// Copyright (C) 2013-2017+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"fmt"
+	errwrap "github.com/pkg/errors"
+	"log"
+	"os/user"
+)
+
+func init() {
+	RegisterResource("user", func() Res { return &UserRes{} })
+}
+
+//TODO
+type UserRes struct {
+	BaseRes `yaml:",inline"`
+	Gid     string `yaml:"gid"`
+	HomeDir string `yaml:"homedir"`
+	State   string `yaml:"state"` // state: exists, absent
+	Uid     string `yaml:"uid"`
+	exec    *ExecRes
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *UserRes) Default() Res {
+	return &UserRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *UserRes) Validate() error {
+	if obj.State != "exists" && obj.State != "absent" {
+		return fmt.Errorf("State must be 'exists' or 'absent'")
+	}
+	return obj.BaseRes.Validate()
+}
+
+// Init initializes the resource.
+func (obj *UserRes) Init() error {
+	obj.exec = &ExecRes{}
+	obj.exec.Init()
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *UserRes) Watch() error {
+
+	// notify engine that we're running
+	if err := obj.Running(); err != nil {
+		return err // bubble up a NACK...
+	}
+
+	var send = false // send event?
+	var exit *error
+	for {
+		select {
+		// NOTE: this part is very similar to the file resource code
+		case event := <-obj.Events():
+			// we avoid sending events on unpause
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit // exit
+			}
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// CheckApply method for User resource. Does nothing, returns happy!
+func (obj *UserRes) CheckApply(apply bool) (checkOK bool, err error) {
+	log.Printf("%s: CheckApply(%t)", obj, apply)
+
+	var exists = true
+
+	usr, err := user.Lookup(obj.GetName())
+
+	log.Printf("%+v", usr)
+
+	if err != nil {
```

Comment:
Not entirely true as per the docs:

>  Lookup looks up a user by username. If the user cannot be found, the returned error is of type UnknownUserError. 

So you only want to do that if the error is that type. Otherwise for "regular" errors we errwrap return an error!